### PR TITLE
Library support in new api (New API PR 3)

### LINKED ISSUE #209: support `m.library` in new api
Bring back an equivalent of:

```ts
const safeMath = m.library("SafeMath");
const contract = m.contract("Contract", {
  libraries: {
    SafeMath: safeMath,
  },
});
```

A placeholder test file stub is here: https://github.com/NomicFoundation/ignition/blob/d7264139e35a1653de591607d50b25fb757d4a0f/packages/core/test/new-api/library.ts#L7

## TODO

* Add `m.library` to new api (follow pattern used for `contract`, and bring across any relevant options and arguments.
* Support `id` provided by the user
* Add simple tests to pin the api usage
* Add `{libraries: { [string]: LibraryFuture }` type object to contract options

This is an api only task, so we are looking to construct the intermediary representation of the deployment. The later phases (i.e. validation, execution etc) are _not_ part of this task.

### LINKED ISSUE #210: libraryFromArtifact
The library equivalent of `m.contractFromArtifact`:

```ts
const library1 = m.contractFromArtifact("Library1", fakeArtifact, [
  1,
  "a",
  BigInt("9007199254740991"),
]);

return { contract1: library1 };
```

A placeholder test file stub is here: https://github.com/NomicFoundation/ignition/blob/d7264139e35a1653de591607d50b25fb757d4a0f/packages/core/test/new-api/libraryFromArtifact.ts#L9

## TODO

* Add `m.contractFromArtifact` to new api (follow pattern used for `contractFromArtifact`, and bring across any relevant options and arguments
* Support `id` provided by the user
* Add simple tests to pin the api usage

This is an api only task, so we are looking to construct the intermediary representation of the deployment. The later phases (i.e. validation, execution etc) are _not_ part of this task.

### PR DESCRIPTION
**Pull request by zoeyTM**
fixes #209 
fixes #210

### CODE REVIEW

File: packages/core/src/new-api/internal/module-builder.ts [hunk 1]
```diff
@@ -101,6 +113,56 @@ export class IgnitionModuleBuilderImplementation<
...
   }
 
+  public library<LibraryNameT extends string>(
+    libraryName: LibraryNameT,
+    options: LibraryOptions = {}
+  ): NamedLibraryDeploymentFuture<LibraryNameT> {
+    const id = options.id ?? libraryName;
+    const futureId = `${this._module.id}:${id}`;
+
+    this._assertUniqueContractId(futureId);
```
Comment (kanej) [inline packages/core/src/new-api/internal/module-builder.ts:123]: Lets use a library specific version here, then clean up the duplication in a later PR.

File: packages/core/src/new-api/internal/module-builder.ts [hunk 2]
```diff
@@ -64,6 +70,12 @@ export class IgnitionModuleBuilderImplementation<
       future.dependencies.add(afterFuture);
     }
 
+    for (const libraryFuture of Object.values(options.libraries ?? {}).filter(
```
[the remark this answers is not in the document]
Comment (zoeyTM) [reply packages/core/src/new-api/internal/module-builder.ts:73]: It's unclear what library linking will look like later, so I think we punt on this for now. It's easy to add later once we have more info

**Comment by alcuadrado** [comment]
I think `ContractFromArtifactOptions` should also include `libraries`, as artifacts have libs.

File: packages/core/src/new-api/internal/module-builder.ts (hunk 1 above)
Comment (alcuadrado) [reply packages/core/src/new-api/internal/module-builder.ts]: On the contrary, future ids are what are unique, it doesn't matter if they come from different type of futures.

For example, any two of those being present should lead to an error:

```ts
const c = m.contract("C");
const c2 = m.library("C");
const c3 = m.library("C3", {id: "C"});
```

Maybe John wanted to make the suggestion more specific?

I think I would go with an error message like 

`Duplicated id ${id} found in module ${moduleId}. Please fix this by passing an explicit id to the second future that's using it.`

File: packages/core/src/new-api/internal/module-builder.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (alcuadrado) [reply packages/core/src/new-api/internal/module-builder.ts:130]: We need to keep track of this info in a more explicit way in the different deployment futures.

File: packages/core/src/new-api/internal/module-builder.ts (hunk 1 above)
Comment (kanej) [reply packages/core/src/new-api/internal/module-builder.ts]: The id check is generic, every future id is compared with every other future id, whatever the type of future.

But the check needs parameterised in to variables:
1. The error message displayed `Your library has an id that has already been used` or some such
2. the function the check is being called from to allow filtering on the stack trace

As @alcuadrado mentions the message can be made generic, at that point you could just pass the filtering function, so maybe:

```
this._assertUniqueId(futureId, this.contract)
```

File: packages/core/src/new-api/internal/module-builder.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (kanej) [reply packages/core/src/new-api/internal/module-builder.ts:130]: Store the object of labels -> contract futures inside the deployment futures:

* contract
* contractFromArtifact
* library
* libraryFromArtifact

**Review by kanej — APPROVED** [review]
Looks good.

**Pull request merged**